internal/queue/queue.go

```diff
@@ -247,3 +280,104 @@ func NewQueue(discovered discovery.DiscoveredConfigs) (*Queue, error) {
 
 	return q, errors.New("cycle detected during queue construction")
 }
+
+// GetReadyWithDependencies returns all entries that are ready to run and have all dependencies completed(or no dependencies).
+func (q *Queue) GetReadyWithDependencies() []*Entry {
+	out := make([]*Entry, 0, len(q.Entries))
+
+	for _, e := range q.Entries {
+		if e.Status == StatusReady {
+			allDepsReady := true
+
+			for _, dep := range e.Config.Dependencies {
+				depEntry, ok := q.Index[dep.Path]
+				if !ok || depEntry.Status != StatusSucceeded {
+					allDepsReady = false
+					break
+				}
+			}
+
+			if allDepsReady {
+				out = append(out, e)
+			}
+		}
+	}
+
+	return out
+}
+
+// SetStatus sets the status of the given entry, handles fail-fast logic if enabled, and updates the statuses of dependents accordingly.
+// This method should be used for all status transitions. If fail-fast is enabled and the entry fails, all not-yet-started tasks are marked as failed.
+// It also unblocks children or marks them as failed/ancestor-failed as appropriate.
```

Comment:
NIT: Consider whether it's better to have that be a different method. Something like "PropagateFailure". Maybe an option on a `StatusSetter` or something. Nothing about the method signature here indicates that these side-effects are going to occur. You have to read the comment to know that they'll happen.